---
title: "Untitled"
author: "Shaleigh Smith"
date: "4/4/2019"
output: html_document
---

Shaleigh Smith
NGS Coursework #6

---

```{r}

# Import libraries
library(tidyverse)
library(exomePeak)
library(dplyr)

```


---

```{r}

# Set working directory
setwd("/Users/sha/Desktop/NGS_Informatics/NGS_courswork/ngs_coursework6_shaleigh_smith")

```



---

```{r}

# Set paths 

gtf <- "./genes.gtf"

# Input RNA
i_control_1 <- "./iCTRL1_sorted.bam"
i_control_2 <- "./iCTRL2_sorted.bam"
i_control_3 <- "./iCTRL3_sorted.bam"
i_sample_1 <- "./iDS1_sorted.bam"
i_sample_2 <- "./iDS2_sorted.bam"
i_sample_3 <- "./iDS3_sorted.bam"

# m6a Enriched RNA (treated)
m_control_1 <- "./mCTRL1_sorted.bam"
m_control_2 <- "./mCTRL2_sorted.bam"
m_control_3 <- "./mCTRL3_sorted.bam"
m_sample_1 <- "./mDS1_sorted.bam"
m_sample_2 <- "./mDS2_sorted.bam"
m_sample_3 <- "./mDS3_sorted.bam"

```



---

```{r}

result <- exomepeak(GENE_ANNO_GTF = gtf,
                    IP_BAM = c(m_control_1, 
                               m_control_2, 
                               m_control_3),
                    INPUT_BAM = c(i_control_1, 
                                  i_control_2, 
                                  i_control_3),
                    TREATED_IP_BAM = c(m_sample_1, 
                                       m_sample_2, 
                                       m_sample_3),
                    TREATED_INPUT_BAM = c(i_sample_1, 
                                          i_sample_2, 
                                          i_sample_3))

```



---

```{r}

# View results
con_sig_diff_peak <- read.table("./exomePeak_output/con_sig_diff_peak.xls", head = TRUE,
                                sep = "\t")

sig_diff_peak <- read.table("./exomePeak_output/sig_diff_peak.xls", head = TRUE,
                                sep = "\t")

diff_peak <- read.table("./exomePeak_output/diff_peak.xls", head = TRUE,
                                sep = "\t")

as.tibble(con_sig_diff_peak)
as.tibble(sig_diff_peak)
as.tibble(diff_peak)

```

```{r}

# Modified Transcripts that differ between test and control datasets
nrow(con_sig_diff_peak) # 447

# Number of genes modified in each individual dataset

mcols(tmp_rs)

```



---


---
